File: problems/chapter11/nicetak/arc065_b.jl
struct UnionFind
    par::Vector{Int}
    size::Vector{Int}
    UnionFind(n) = new(collect(1:n), ones(Int, n))
end
 
root!(uf, x) = uf.par[x] == x ? x : uf.par[x] = root!(uf, uf.par[x])
 
same(uf, x, y) = root!(uf, x) == root!(uf, y)        
 
function unite!(uf, x, y)
    x = root!(uf, x)
    y = root!(uf, y)
    x == y && return false
 
    if uf.size[x] > uf.size[y]
        uf.par[y] = x
        uf.size[x] += uf.size[y]
    else
        uf.par[x] = y
        uf.size[y] += uf.size[x]
    end
    return true
end
size!(uf, x) = uf.size[root!(uf, x)]

function main()
    N, K, L = parse.(Int, split(readline()))

    uf₁ = UnionFind(N)
    uf₂ = UnionFind(N)

    for _ in 1:K
        p, q = parse.(Int, split(readline()))
        unite!(uf₁, p, q)
    end

    for _ in 1:L
        r, s = parse.(Int, split(readline()))
        unite!(uf₂, r, s)
    end

    d = Dict{Tuple{Int, Int}, Int}()
    for i in 1:N
        key = (root!(uf₁, i), root!(uf₂, i)) 
        if haskey(d, key)
            d[key] += 1
        else
            d[key] = 1
        end
    end

    for i in 1:N-1
        print(d[(root!(uf₁, i), root!(uf₂, i))], " ")
    end
    
    println(d[(root!(uf₁, N), root!(uf₂, N))])
end

main()

File: problems/chapter11/Ysi/arc065_b.py
class UnionFind():
    def __init__(self, n) -> None:
        self.n = n
        self.par = [-1] * n
        self.siz = [1] * n

    def root(self, x):
        if self.par[x] == -1:
            return x
        else:
            self.par[x] = self.root(self.par[x])
            return self.par[x]

    def issame(self, x, y):
        return self.root(x) == self.root(y)

    def unite(self, x, y):
        x = self.root(x)
        y = self.root(y)

        if x == y:
            return False

        if self.siz[x] < self.siz[y]:
            x, y = y, x

        self.par[y] = x
        self.siz[x] += self.siz[y]
        return True

    def size(self, x):
        return self.siz[self.root(x)]
    
    def members(self, x):
        return [i for i in range(self.n) if self.issame(x, i)]

def main():
    N, K, L = map(int, input().split())
    load = []
    for _ in range(K):
        p, q = map(int, input().split())
        load.append([p-1, q-1])
    railway = []
    for _ in range(L):
        r, s = map(int, input().split())
        railway.append([r-1, s-1])
    
    ufL = UnionFind(N)
    ufR = UnionFind(N)

    for e in load:
        a, b = e
        ufL.unite(a, b)
    for e in railway:
        a, b = e
        ufR.unite(a, b)

    
    res = dict()
    roots = []
    for i in range(N):
        root = (ufL.root(i), ufR.root(i))
        roots.append(root)
        if root in res:
            res[root] += 1
        else:
            res[root] = 1
    
    ans = []
    for i in range(N):
        root = roots[i]
        ans.append(res[root])

    print(*ans)
if __name__=='__main__':
    main()